src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -10,6 +10,76 @@
 #include "mpidu_init_shm.h"
 #include "xpmem_seg.h"
 
+MPL_STATIC_INLINE_PREFIX int init_coop_counter(int local_rank, int num_local,
+                                               size_t page_sz,
+                                               OPA_int_t ** coop_counter_align,
+                                               void **coop_counter_origin)
+{
+    int mpi_errno = MPI_SUCCESS, i;
+    OPA_int_t *local_cc_align = NULL;
+    void *local_cc_origin = NULL;
+    uint64_t rank0_cc_addr;
+    size_t cc_sz = num_local * num_local * sizeof(OPA_int_t);
+
```

Comment:
Last P1/T1 receives data from `src=0`, is this example correct? 
counter array is two-dimensional. if both P0 and P1 send RTS to P1, P1 won't use the same counter to receive these two packets. As long as the upper layer makes sure only one thread is using this queue, our code should be safe.

Reply:
Two threads may call MPI concurrently. We cannot assume that the upper layer makes critical section for the entire send/recv path. Rohit is working on per-vci optimization, which will allow two threads touch SHM concurrently. Thus, you need to ensure the queue logic itself is thread-safe.

Reply:
logic is correct to work on multithreads environment.
As long as we add our lock to dmessage_queue, it should be safe

Reply:
It is correct if we add critical section when touching the dmsg queue. But the performance will be bad, because it allow sonly one thread on a process to copy the message at a time. In my example, `P1/T0` and `P1/T1` cannot copy concurrently. The trend of MPICH is to minimize serialization of multithreading messages, thus I do not think the design will work for long term. I'd change it before merging into master.